### PR TITLE
New PR to replace fix981_canonAutoFocus2 which will be closed.

### PR DESCRIPTION
**Pull request by clanmills**
This PR replaces fix981_canonAutoFocus2 (with the same code).  This has been done because there was a long discussion about fix981_canonAutoFocus2.  For easier team review, there is no need to read about 30 discussion items in fix981_canonAutoFocus2

fix981_canonAutoFocus2 was a replacement for fix981_canonAf which was abandoned as it revealed architectural limits in the binary tag decoders.

This PR (and fix981_canonAutoFocus2) use a decoder listed in TiffMappingInfo to decode Exif.Canon.AFInfo.   The decoding function "manufactures" Exif tags such as Exif.Canon.AFNumPoints from the data in Exif.Canon.AFInfo.  These tags must never be written to file and are removed from the metadata in exif.cpp/ExifParser::encode().

Three of the tags created (AFPointsInFocus,AFPointsSelected, AFPrimaryPoint) are bitmasks.  As the camera can have up to 64 focus points, the tags are a 64 bit mask to say which points are active.  The function printBitmask() reports data such as `1,2,3` or `(none)`.

This decoding function decodeCanonAFInfo() added to TiffMappingInfo manufactures the new tags.  Normally, tags are processed by the binary tag decoder and that approach was taken in branch fix981_canonAf.  However, the binary tag decoder cannot deal with AFInfo because the size of some metadata arrays cannot be determined at compile time.

We should support decoding AFInfo in 0.28, however we should NOT auto-port this PR.  We can avoid having to explicitly delete tags from the metadata before writing by adding a "read-only" flag to TagInfo.  This would break the Exiv2 v0.27 API and has been avoided.  There is an array in decodeCanonAFInfo() which lists the "manufactured" tags such as Exif.Canon.AFNumPoints.  In the Exiv2 v0.28 architecture, a way might be designed to generate that data at run-time.

It has been interesting to work on this issue.  I've learnt more about the Exiv2 architecture for tag and MakerNote decoding and these lessons will be documented in the book being written for #911 .

### CODE REVIEW

**Review by derselbst — COMMENTED** [review]
I gave this a final test and found one minor issue: `Exif.Canon.AFInfoSize` should be added to the filtered tags as well. Apart from that it works absolutely fine. Thank you very much for your effort, Robin!

**Comment by clanmills** [comment]
Good Catch, Tom.  I missed that.  I cut'n'pasted those keys from another file and dropped one.  Thanks for spotting and reporting that.

I'm not a photographer in the sense of knowing how about cameras really work.  For my Nikon D5300, in manual focus mode, there is a red rectangle which you can move about with the arrow keys.  Presumably the "PrimaryPoint" bitmask tells you which points lie inside. 

From my school-boy Physics, I thought a lens has a focal length and the aperture determines the depth of field.  So, my guess is that the camera selects a point as its "Primary Focus".  From radar (or ultra-sound), it knows the points in the image that seem to be in focus and records that in Exif.Canon.AFPointsInFocus.    

My neighbour is a member of the local photography club.  Maybe they'll ask me to give a talk about Exiv2 and I could ask them to explain some of these puzzles.

File: src/tags_int.cpp
Comment (D4N) [inline src/tags_int.cpp]: What's this change here?

File: src/tags_int.cpp
Comment (D4N) [inline src/tags_int.cpp]:
```suggestion
                uint16_t bits = static_cast<uint16_t>(value.toLong(i));
```

File: tests/bugfixes/github/test_issue_981.py
Comment (D4N) [inline tests/bugfixes/github/test_issue_981.py]: Can you please move the commas to the end of the preceding lines? I usually don't like to argue about style & formatting, but that is against every Python style guide that ever existed.

File: src/tags_int.hpp
Comment (D4N) [inline src/tags_int.hpp]: Could you add an example output or two?

File: tests/bugfixes/github/test_issue_981.py
Comment (D4N) [inline tests/bugfixes/github/test_issue_981.py]:
```suggestion
    stderr = [""] * 6
```

File: tests/bugfixes/github/test_issue_981.py
Comment (D4N) [inline tests/bugfixes/github/test_issue_981.py]:
```suggestion
    retval = [0] * 6
```

File: src/tiffvisitor_int.cpp
Comment (D4N) [inline src/tiffvisitor_int.cpp]: There is no guarantee that `uint` contains at least 3 elements.
```suggestion
        const uint16_t nPoints = uint.at(2);
```

File: src/tiffvisitor_int.cpp
Comment (D4N) [inline src/tiffvisitor_int.cpp]:
```suggestion
                    for ( int16_t k = 0 ; k < records[i].size ; k++ ) s << " " << ints.at(nStart++);
```

File: src/tiffvisitor_int.cpp
Comment (D4N) [inline src/tiffvisitor_int.cpp]:
```suggestion
                    for ( int16_t k = 0 ; k < records[i].size ; k++ ) s << " " << uint.at(nStart++);
```

File: src/tiffvisitor_int.cpp
Comment (D4N) [inline src/tiffvisitor_int.cpp]: Isn't this equivalent to:
```suggestion
        const uint16_t count = 8 + 4 * nPoints + 3 * nMasks;
```

**Review by D4N — CHANGES_REQUESTED** [review]
Overall this looks good, I've flagged a few potential issues.

File: src/tags_int.cpp
Comment (clanmills) [reply src/tags_int.cpp]: I detabbed the file.

File: tests/bugfixes/github/test_issue_981.py
Comment (clanmills) [reply tests/bugfixes/github/test_issue_981.py]: I'm not going to argue.

File: src/tags_int.cpp
Comment (clanmills) [reply src/tags_int.cpp]: Changed.

File: src/tags_int.hpp
Comment (clanmills) [reply src/tags_int.hpp]: Done.

File: tests/bugfixes/github/test_issue_981.py
Comment (clanmills) [reply tests/bugfixes/github/test_issue_981.py]: Good Idea!  Done as `*len(commands)`

File: src/tiffvisitor_int.cpp
Comment (clanmills) [reply src/tiffvisitor_int.cpp]: Modified the `insufficient data` test on line 475 (great catch, well done Dan).
Adopted `.at(2)`.

File: src/tiffvisitor_int.cpp
Comment (clanmills) [reply src/tiffvisitor_int.cpp]: Adopted `.at[nStart++]`

File: src/tiffvisitor_int.cpp
Comment (clanmills) [reply src/tiffvisitor_int.cpp]: Well yes, and no.  It's not static.  At the moment it's 4 records of nPoints and 3 of nMasks.  However In development, we've modified the number of records with nPoints and nMasks in the `records[]` array more than once.  It's safer to count how much data we need, and be certain it's available.

**Review by D4N — DISMISSED** [review]

**Review by D4N — APPROVED** [review]
@clanmills Thank you for addressing my comments! I've squashed the commits and rebased them on 0.27-maintenance and would merge this once the CI is green again.

**Comment by derselbst** [comment]
Pls. consider to rename `AFPrimaryPoints` based on my research, before merging this:

https://github.com/Exiv2/exiv2/issues/981#issuecomment-524527751 :

> [...] unusable AF points are reported as AFPrimaryPoints.
>
> Perhaps it would be worth to rename that field to something like AFPointsUnusable or AFPointsDisabled.

**Comment by clanmills** [comment]
@derselbst I thought we adopted "Primary" to align with the Exiftool documentation.  https://sno.phy.queensu.ca/~phil/exiftool/TagNames/Canon.html

I'm OK with changing this, however perhaps you could discuss with Phil.  Would it be OK to accept things as they stand and I can get Exiv2 v0.27.3 finished.  We can change the name as a separate issue in v0.27.4 if ExifTool adopts your suggestion.  OK?

I hope to get Exiv2 v0.27.3 RC1 released early next week and (if nothing bad happens), that code will be released as v0.27.3 on 2019-09-30 (on schedule).

Exiv2 v0.27.4 is scheduled for 2019-12-31.

**Comment by derselbst** [comment]
I just wanted to make sure that we

* don't forget about this, and
* don't unnecessarily break API between releases by changing names.

On the other hand, I don't mind if you guys want to postpone renaming to the next maintenance release. In this case, I would just open a new issue to discuss this.

**Comment by clanmills** [comment]
I don't believe the name is revealed via the API.  I think we should open a new issue to deal with your concern and I will ask Phil Harvey to comment.

**Pull request merged**
